### PR TITLE
Timeout in msecs

### LINKED ISSUE #291: use milliseconds internally, and don't make rounding errors
Broken out from closed issue #276 

@knopwob wrote:

Eventhough the protocol works with milliseconds, dunst internally uses seconds, so the timeout send by the app will be rounded to the next second.

so notify-send -t 100 "blehp" should timeout after one second. This might get extended to 2 seconds, due to some other rounding stuff.

@quite wrote:

For me, dunst git as of today lets the notification sit around for milliseconds rounded up to whole second(s) plus 1 second. As @knopwob lays it out.

### PR DESCRIPTION
**Pull request by bebehei**
Timeout in milliseconds and do not make rounding errors. Additionally use the monotonic time instead of `time(NULL)`, which is vulnerable to timeshifts.

To do:

- <del>add an option `option_get_time`, which allows timespans of `1ms`, `1s`, `1m`  to have variable timeout parameters (currently this is quickfixed via hardcoded `[...] = 1000 * 1000 * option_get_int( [...]`  </del> Done.
- <del>what's the workload, when x_is_idle and timeout = 1ms of notification?</del>
    - <del>obviously, dunst is not that fast. I think we should add a minimal threshold</del>
- <del>bug: show age_threshold time does not increment in evenly spaced values.</del>
    - really solved, but the age_threshold is always a second behind. The problem is somewhere else in `notification_update_text_to_render`.

Fixes #291

### CODE REVIEW

**Comment by bebehei** [comment]
@tsipinakis I'd like to have a review here. It works now perfectly on my own system.

**Comment by tsipinakis** [comment]
Title is `Timeout in msecs`

But you internally implemented timeout in microseconds, was that intentional?

File: docs/dunst.pod [hunk 1]
```diff
@@ -237,6 +237,11 @@ Set to 0 to disable.
 Transient notifications will ignore this setting and timeout anyway.
 Use a rule overwriting with 'set_transient = no' to disable this behavior.
 
+=item B<min_timeout> (default: 0)
```
[the remark this answers is not in the document]
Comment (bebehei) [reply docs/dunst.pod]: There is no "sane" value. I think it would be best to default it to 1, as this was dunst's previous behavior.

File: src/dunst.c [hunk 2]
```diff
@@ -173,61 +172,39 @@ void wake_up(void)
...
+                }
         }
 
-        /* show_age_timeout might be negative */
-        if (min_timeout < 1) {
-                return 1;
-        } else {
-                return min_timeout;
-        }
+        return sleep;
```
[the remark this answers is not in the document]
Comment (bebehei) [reply src/dunst.c]: > If we have nothing to do, there is no reason to add a timer. Best solution is probably to return something like -1.

This function is only called, if notifications are available -> there has to be something returned, which represents a timeout. You'll never call this function if you don't have to sleep.

> PS: With this, I just realised that the result of g_get_monotonic_time is a signed int, I wonder why?

It's simply error checking. If you've got an overflow, you can detect it, by geting a negative value. Additionally, I like it as it's easier to express negative timedifferences.

File: src/dunst.c [hunk 3]
```diff
@@ -244,13 +221,13 @@ gboolean run(void *data)
         }
 
         if (xctx.visible) {
-                int now = time(NULL);
-                int sleep = get_sleep_time();
+                gint64 now = g_get_monotonic_time();
+                gint64 sleep = get_sleep_time();
+                gint64 timeout_at = now + sleep;
 
                 if (sleep > 0) {
```
[the remark this answers is not in the document]
Comment (bebehei) [reply src/dunst.c]: Full ACK.

File: src/notification.c [hunk 4]
```diff
@@ -635,26 +636,26 @@ void notification_update_text_to_render(notification *n)
         }
 
         /* print age */
-        int hours, minutes, seconds;
-        time_t t_delta = time(NULL) - n->timestamp;
+        gint64 hours, minutes, seconds;
```
[the remark this answers is not in the document]
Comment (bebehei) [reply src/notification.c]: "everything gint64".

File: src/option_parser.c [hunk 5]
```diff
@@ -123,6 +123,19 @@ char *ini_get_string(char *section, char *key, const char *def)
...
 }
 
+gint64 ini_get_time(char *section, char *key, gint64 def)
+{
+        char *timestring = get_value(section, key);
+        gint64 val = def;
+
+        if (timestring) {
+                val = string_to_time(timestring);
+                g_free(timestring);
```
[the remark this answers is not in the document]
Comment (bebehei) [reply src/option_parser.c]: 🤦‍

File: src/option_parser.c [hunk 6]
```diff
@@ -443,6 +470,17 @@ char *option_get_string(char *ini_section, char *ini_key, char *cmdline_key,
 
 }
 
+gint64 option_get_time(char *ini_section, char *ini_key, char *cmdline_key,
+                       gint64 def, char *description)
+{
+        /* *str is only used to check wether the cmdline option is actually set. */
+        char *str = cmdline_get_value(cmdline_key);
+        g_free(str);
```
[the remark this answers is not in the document]
Comment (bebehei) [reply src/option_parser.c]: LOL, these 3 lines are the most stupid code, I've ever passed in for a PR.

File: docs/dunst.pod (hunk 1 above)
[the remark this answers is not in the document]
Comment (tsipinakis) [reply docs/dunst.pod]: > There is no "sane" value

Yeah probably, but I also think there's no clear use for it. If a program sets a abnormal timeout value it's the clients fault not the servers (and can be overridden via rules anyway).

File: src/dunst.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (tsipinakis) [reply src/dunst.c]: > This function is only called, if notifications are available -> there has to be something returned, which represents a timeout

Not if the notification has no timeout (which also reminds me, that's also a case not handled in your implementation).

File: src/dunst.c [hunk 7]
```diff
@@ -173,61 +172,39 @@ void wake_up(void)
...
-                        if (have_ttl) {
-                                min_ttl = MIN(min_ttl, ttl);
-                        } else {
-                                min_ttl = ttl;
-                                have_ttl = true;
-                        }
-                }
-        }
+                if (ttl >= 0)
+                        sleep = MIN(sleep, ttl);
```
[the remark this answers is not in the document]
Comment (tsipinakis) [reply src/dunst.c]: Since it came up in another comment: You should also handle separately the case of `n->timeout <= 0`, since that will also generate a negative `ttl` but for a different reason (notification doesn't expire).

File: docs/dunst.pod (hunk 1 above)
[the remark this answers is not in the document]
Comment (bebehei) [reply docs/dunst.pod]: > (and can be overridden via rules anyway).

Matching all notifications below 1000ms with a rule is quite hard IMHO. That's the  reason, why I'd like to have it.

File: src/option_parser.c (hunk 5 above)
[the remark this answers is not in the document]
Comment (bebehei) [reply src/option_parser.c]: See #386

File: src/utils.c [hunk 8]
```diff
@@ -118,6 +118,24 @@ char *string_to_path(char *string) {
...
+        else if (strstr(string, "s"))
+                factor = (gint64)G_USEC_PER_SEC;
+        else if (strstr(string, "m"))
+                factor = (gint64)G_USEC_PER_SEC * 60;
+        else if (strstr(string, "h"))
+                factor = (gint64)G_USEC_PER_SEC * 60 * 60;
+        else if (strstr(string, "d"))
+                factor = (gint64)G_USEC_PER_SEC * 60 * 60 * 24;
+
+        return strtol(string, NULL, 10) * factor;
```
[the remark this answers is not in the document]
Comment (bebehei) [reply src/utils.c]: I see, that checking for `ERANGE` is a smart decision, but I don't see any sense checking for`EINVAL`. Even the [manpage did not include it in the example](https://manpages.debian.org/stretch/manpages-dev/strtol.3.en.html#ERRORS).

**Comment by bebehei** [comment]
Note to myself: I should explain the timeformat in the docs and change the docs for any threshold/timeout value according to the timeformat.

File: src/utils.c [hunk 9]
```diff
@@ -118,6 +120,41 @@ char *string_to_path(char *string) {
...
+        errno = 0;
+        char *endptr;
+        gint64 val = strtoll(string, &endptr, 10);
+
+        if (errno == ERANGE)
+                fprintf(stderr, "ERROR: Time: '%s' overflows.\n", string);
+        else if (errno == EINVAL)
+                fprintf(stderr, "ERROR: Time: '%s' invalid value.\n", string);
+        else if (string == endptr)
+                fprintf(stderr, "ERROR: Time: No digits found.\n", string);
```
Comment (tsipinakis) [inline src/utils.c]: This pops a warning, `string` is not used inside the format.

**Comment by tsipinakis** [comment]
> There is a weird bug that corrupts the default icon names

I re-tested 97d878c1b67e096a7c470d4c741da6e1f8c77c73 and it's not there (I probably forgot to pull before merging - oops). Bisect shows that it was fixed in 50eb61cc79fa867fd683ad64e6e1e02016610bea - weird.

File: test/utils.c [hunk 10]
```diff
@@ -130,6 +130,20 @@ TEST test_string_to_path(void)
...
 }
 
+TEST test_string_to_time(void)
+{
+        char *input[] = { "5000 ms", "5000ms",  "100", "10s",   "2m",    "11h",      "9d", NULL };
+        gint64  exp[] = {      5000,     5000, 100000, 10000, 120000, 39600000, 777600000 };
+
+        int i = 0;
+        while (input[i]){
+                ASSERT_EQ(string_to_time(input[i]), exp[i]*1000);
```
Comment (tsipinakis) [inline test/utils.c]: From the recent failed test, the error for this assertion is unreadable:
```
FAIL test_string_to_time: string_to_time(input[i]) != exp[i]*1000 (test/utils.c:140)
```

[`ASSERT_EQ_FMT`](https://github.com/silentbicycle/greatest#assert_eq_fmtexpected-actual-format) should be used to explain what was tested and what failed.

Comment (bebehei) [reply test/utils.c]: I'm already in it ;-)

Comment (bebehei) [reply test/utils.c]: But thanks for the hint with `ASSERT_EQ_FMT`. That's gold

**Comment by bebehei** [comment]
Sorry, I accidentally rebased all fixup commits (instead of rebasing the fixup commits) 🙈 . So I took the opportunity to rebase it on master.

> I'm still not a fan of min_timeout. Honestly it feels like something that should be possible to do via rules but our rule implementation is not powerful enough (yet). I'd prefer if this was hardcoded for now. I'm not sure of any other usecases other than overriding ridiculous timeouts set by some applications.

Let's think different about this: If we assume, an application never changes its notificaiton timeout. If this is given, then there is no need to have a `min_timeout`, as you could match it with a rule (even on a per-application basis). This is far better. I removed the `min_timeout` commit.

> This conflicts with #386 and there are a lot of warnings about the const types, perhaps merging this while this PR was in progress was a bad idea :/

It was intended. The whole warnings were printed, because I fixed up `option_get_time` already to have const arguments here in this PR. (And the other functions still misused it).

**Comment by tsipinakis** [comment]
> Sorry, I accidentally rebased all fixup commits (instead of rebasing the fixup commits)

That's fine, I've been away for quite a while so I'll probably start the review from the beginning anyway. For future reference though, `git reflog` can be handy :)

**Comment by bebehei** [comment]
> That's fine, I've been away for a few days so I'll probably start the review from the beginning anyway. For future reference though, git reflog can be handy :)

I know git reflog, but it was even too late for that. I've seen it, when I had already pushed the branch. And I guess, GitHub is not that smart to catch, that it had been the previous state. 🙈

**Comment by tsipinakis** [comment]
Works fine as far as I can test, after the open refactors are done we should focus more on automated testing since it's basically nil.

**Comment by tsipinakis** [comment]
Apparently githubs `resolve conflicts` functionality creates a merge commit into the PR instead of just resolving the conflicts in the actual merge commit to master. Can you remove the latest commit and rebase on top of master?

**Comment by bebehei** [comment]
> automated testing since it's basically nil.

Yes, please! This makes accepting changes much easier!

**Comment by bebehei** [comment]
```
git checkout master && git pull https://github.com/bebehei/dunst testmastermerge && git push origin master
```
